### ISSUE TITLE
Macros leaking between build ids?

### ISSUE THREAD

**Issue by jcreedcmu** [bhauman/lein-figwheel#169]
tl;dr I think I have observed lein-figwheel, when focused on one build-id, erroneously using macro definitions from another build-id.

In more detail, in the repository https://github.com/jcreedcmu/cljsbuild-leak I did:

```shell
lein clean && lein figwheel foo
cat build/foo/js/out/pack/core.js
```

and I saw

```javascript
// Compiled by ClojureScript 0.0-3269 {}
goog.provide('pack.core');
goog.require('cljs.core');
cljs.core.print.call(null,new cljs.core.Keyword(null,"bar","bar",-1386246584));
```

and I was surprised to see the keyword `:bar` being constructed rather
than `:foo`. In `src/both/pack/core.cljs` the macro `env` is invoked,
and I expect (when building build-id `foo`) to use the definition of
that macro in `src/foo/pack/env.clj`, not the one in
`src/bar/pack/env.clj`. Is this a bug?

If I delete the entire build
```clojure
                       :bar {:source-paths ["src/both" "src/bar"]
                             :compiler {:main          "pack.core"
                                        :output-dir    "build/bar/js/out"
                                        :output-to     "build/bar/app.js"}}
```

from `project.clj` and do

```shell
lein clean && lein figwheel foo
cat build/foo/js/out/pack/core.js
```

again, I see

```javascript
// Compiled by ClojureScript 0.0-3269 {}
goog.provide('pack.core');
goog.require('cljs.core');
cljs.core.print.call(null,new cljs.core.Keyword(null,"foo","foo",1268894036));
```

as I expect.

Incidentally, I also get the expected behavior
(`cljs.core.Keyword(null,"foo","foo",...)` emitted) when directly
running ```lein cljsbuild once foo```, which is why I'm provisionally filing the issue here against lein-figwheel.

Any advice or redirection to the appropriate place (or advice to adjust my expectations to be more in line with reality in the event that I'm just naively expecting the wrong thing) much appreciated.

**Comment by bhauman** [comment]
First of all thanks for creating a minimal example.

This is interesting behavior.

It's really not correct for a for an individual build.  But it may be a
moot because if you are going to be working on foo and bar side by side as
in:

$ lein figwheel foo bar

Then this is going to be a big problem anyway. This is more than likely not
the best strategy to do conditional inclusion stuff.

I am going to look and find out why the other source-paths are initially
being loaded anyway.

Thanks again for making such a great example.  But again, be wary of going
down this path. If you do a side by side build in cljsbuild (ie. lien
cljsbuild auto foo bar ) you are likely going to run into similar problems.

Bruce

**Comment by jcreedcmu** [comment]
Ok, thanks for the warning! *Is* there a good strategy for doing any kind of conditional inclusion/compilation, e.g. two simultaneous builds where the "real" one calls into a real ajax library and the "test" one calls into some mocks? Is this specific to lein-figwheel's treatment of .clj, i.e. will I run into the same problems with mere .cljs files defining the same namespace spanning the two builds? I could probably live with a simple single runtime test against whether-I'm-in-test-env-or-not for each call into the ajax library. (but of course I went down this road trying to avoid exactly that :) )

**Comment by bhauman** [comment]
You can do conditional .cljs very easily. It's the clj that's causing the problem.
